### PR TITLE
sasarkar/Fix recompiles if limit_hpu_graph is False

### PR DESCRIPTION
**Pull request by ssarkar2**
# What does this PR do?
**TL;DR**
Fix recompiles if limit_hpu_graph is False

**Details**
If `--limit_hpu_graphs` is set to False, there are multiple recompilations, causing low thruput.

The reason for recompilations is:
for 1st prefill, input = prompt of len “p” goes in, and kv cache of size “p” comes out, and that is padded by “m” to give size = p+m
for second prefill (during second warmup), inp = p goes in, but output is p+m (should have been just p).. which is padded by m again, and then becomes p+m+m etc
 This is probably because this clearing code isnt invoked when limit_hpu_graph=False: [optimum-habana/optimum/habana/transformers/generation/utils.py at main · huggingface/optimum-habana](https://github.com/huggingface/optimum-habana/blob/main/optimum/habana/transformers/generation/utils.py#L1888)

A possible fix is to check if inputsize==output_kv_size, and pad only in that case, which is what this PR does. Note that if input sizes are different in each step, this might not work correctly (but that is not the case in normal mode of operations of text gen using run_generation)

**Tests/Experiments**
basic cmd line. This is a toy case with small inp/outs
```
QUANT_CONFIG=./quantization_config/maxabs_quant.json TQDM_DISABLE=1 python run_generation.py --model_name_or_path meta-llama/Llama-2-7b-hf --attn_softmax_bf16 --use_hpu_graphs --trim_logits --warmup 2 --use_kv_cache --bucket_size=4 --bucket_internal --max_input_tokens 8 --max_new_tokens 16 --bf16 --batch_size 1 --use_flash_attention --flash_attention_recompute --n_iter 3
```
Note that when we say limit_hpu_graph=True, we might have to comment out these lines, which set it to false:
https://github.com/huggingface/optimum-habana/blob/main/examples/text-generation/utils.py#L484

| Branch | model | limit_hpu_graph | generation mode | dtype | thruput | 1st tkn latency | 
| -------- | ------- | ------- | ------- | ------- | ------- |------- |
| main | llama7 | False        | greedy | fp8 |  **1.131** | 6.7 - 7.3 |
|  this branch |  llama7 | False    |  greedy | fp8 | **221.78** | 6.6 |
|  main  | llama7 |  True      |  greedy | fp8 | 120.89 | 62-64 |
|  this branch |  llama7 |  True |  greedy | fp8 | 131.36  | 52-54 |
| main |  llama7 | False        | --do_sample | fp8 | **1.12** | 8.8 |
|  this branch | llama7 |  False    |  --do_sample | fp8 | **187.30** | 7.5-8.7 |
|  main  |  llama7 | True      |  --do_sample | fp8 | 117.48 | 54-57 |
|  this branch | llama7 |   True |   --do_sample |fp8 |  114.89  | 57-59 |
| main |  llama7 | False        | greedy | bf16 |  **19.454** | 9.2 |
|  this branch |  llama7 | False    |  greedy | bf16 |**140.02** | 9.1 |
|  main  |  llama7 | True      |  greedy | bf16 | 113.37 | 33 |
|  this branch |   llama7 | True |  greedy | bf16 | 113.5  | 33 |
| main |  falcon7 | False        | greedy | bf16 |  **2.94** | 10-11 |
|  this branch |  falcon7 | False    |  greedy | bf16 | **129.78** | 9.8 |
|  main  |  falcon7 | True      |  greedy | bf16 | 101.67 | 39-43 |
|  this branch |   falcon7 | True |  greedy | bf16 | 102.9 | 39 |

Some other tests:
1. Without `--bucket_size 4 --bucket_internal`, limit_hpu_graph=False
222.52 tps, 6.5 ms FTL ... same as main
2. Without `--bucket_size 4 --bucket_internal`, limit_hpu_graph=True
131.73, 55 ms FTL ... same as main
3. Without `--bucket_internal`, with `--bucket_size 4`, limit_hpu_graph=False
213 tps, 6.2 ms FTL ..same as main
4. Without `--bucket_internal`, with `--bucket_size 4`, limit_hpu_graph=True
131.07 tps, 55 ms FTL ... same as main

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Review by regisss — APPROVED** [review]
LGTM!
